pkg/manifestcomparators/comp_no_new_required_fields.go

```diff
@@ -0,0 +1,142 @@
+package manifestcomparators
+
+import (
+	"fmt"
+	"strings"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+type noNewRequiredFields struct{}
+
+func NoNewRequiredFields() CRDComparator {
+	return noNewRequiredFields{}
+}
+
+func (noNewRequiredFields) Name() string {
+	return "NoNewRequiredFields"
+}
+
+func (noNewRequiredFields) WhyItMatters() string {
+	return "If new fields are required, then old clients will not function properly.  Even if CRD defaulting is used, " +
+		"CRD defaulting requires allowing an object with an empty or missing value to then get defaulted."
+}
+
+func (b noNewRequiredFields) Compare(existingCRD, newCRD *apiextensionsv1.CustomResourceDefinition) (ComparisonResults, error) {
+	if existingCRD == nil {
+		return ComparisonResults{
+			Name:         b.Name(),
+			WhyItMatters: b.WhyItMatters(),
+
+			Errors:   nil,
+			Warnings: nil,
+			Infos:    nil,
+		}, nil
+	}
+	errsToReport := []string{}
+
+	for _, newVersion := range newCRD.Spec.Versions {
+
+		existingVersion := GetVersionByName(existingCRD, newVersion.Name)
+		if existingVersion == nil {
+			continue
+		}
+
+		existingRequiredFields := map[string]sets.String{}
+		existingToSimpleLocation := map[*apiextensionsv1.JSONSchemaProps]*field.Path{}
+		SchemaHas(existingVersion.Schema.OpenAPIV3Schema, field.NewPath("^"), field.NewPath("^"), nil,
+			func(s *apiextensionsv1.JSONSchemaProps, fldPath, simpleLocation *field.Path, _ []*apiextensionsv1.JSONSchemaProps) bool {
+				existingRequiredFields[simpleLocation.String()] = sets.NewString(s.Required...)
+				existingToSimpleLocation[s] = simpleLocation
+				return false
+			})
+
+		// New fields can be required if they are wrapped inside new structs that are themselves optional.
+		// For instance, you cannot add .spec.thingy as required, but if you add .spec.top as optional and at the same
+		// time add .spec.top.thingy as required, this is allowed.
+		// Similar logic exists for adding an array with minlength > 0
```

Comment:
Because of the way that go will serialise this, you need `omitempty` and for `top` to be a pointer in this case. Is it possible to check these conditions if they aren't already being done so?

X ref https://github.com/openshift/enhancements/pull/1411

Reply:
We do not have access to the golang types in the CRD yaml, so we don't have access to the pointer-ness of a field or the omitempty-ness of a field.  This tool could be a reason for adding such markers, but we cannot determine it yet